In [195]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
plt.style.use('seaborn-white')

In [196]:
df = pd.read_csv('customer_supermarket.csv', sep='\t', index_col=0)
df

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0        536365  01/12/10 08:26  2,55     17850.0  United Kingdom  85123A   
1        536365  01/12/10 08:26  3,39     17850.0  United Kingdom   71053   
2        536365  01/12/10 08:26  2,75     17850.0  United Kingdom  84406B   
3        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029G   
4        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029E   
...         ...             ...   ...         ...             ...     ...   
541905   581587  09/12/11 12:50   2,1     12680.0          France   22899   
541906   581587  09/12/11 12:50  4,15     12680.0          France   23254   
541907   581587  09/12/11 12:50  4,15     12680.0          France   23255   
541908   581587  09/12/11 12:50  4,95     12680.0          France   22138   
541909   581587  09/12/11 12:50    18     12680.0          France    POST   

                                  ProdDescr  Qta  
0        WHITE HANGING HEART T-LIGHT HOLDER    6  
1                       WHITE METAL LANTERN    6  
2            CREAM CUPID HEARTS COAT HANGER    8  
3       KNITTED UNION FLAG HOT WATER BOTTLE    6  
4            RED WOOLLY HOTTIE WHITE HEART.    6  
...                                     ...  ...  
541905         CHILDREN'S APRON DOLLY GIRL     6  
541906        CHILDRENS CUTLERY DOLLY GIRL     4  
541907      CHILDRENS CUTLERY CIRCUS PARADE    4  
541908        BAKING SET 9 PIECE RETROSPOT     3  
541909                              POSTAGE    1  

[471910 rows x 8 columns]

In [197]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 471910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         471910 non-null  object 
 1   BasketDate       471910 non-null  object 
 2   Sale             471910 non-null  object 
 3   CustomerID       406830 non-null  float64
 4   CustomerCountry  471910 non-null  object 
 5   ProdID           471910 non-null  object 
 6   ProdDescr        471157 non-null  object 
 7   Qta              471910 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 32.4+ MB


BasketID è una stringa, contiene lettere C che indicano la cancellazione. Eliminiamo le righe corrispondenti. 

In [198]:
import re

def search_letters(the_string):
    if re.search('[a-zA-Z]', the_string) is None:
        return False
    return True

df[df['BasketID'].apply(search_letters)]

BasketID      BasketDate    Sale  CustomerID CustomerCountry  ProdID  \
141     C536379  01/12/10 09:41    27,5     14527.0  United Kingdom       D   
154     C536383  01/12/10 09:49    4,65     15311.0  United Kingdom  35004C   
235     C536391  01/12/10 10:24    1,65     17548.0  United Kingdom   22556   
236     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21984   
237     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21983   
...         ...             ...     ...         ...             ...     ...   
540449  C581490  09/12/11 09:57    0,83     14397.0  United Kingdom   23144   
541541  C581499  09/12/11 10:28  224,69     15498.0  United Kingdom       M   
541715  C581568  09/12/11 11:57   10,95     15311.0  United Kingdom   21258   
541716  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   84978   
541717  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   20979   

                               ProdDescr  Qta  
141                             Discount   -1  
154      SET OF 3 COLOURED  FLYING DUCKS   -1  
235       PLASTERS IN TIN CIRCUS PARADE   -12  
236     PACK OF 12 PINK PAISLEY TISSUES   -24  
237     PACK OF 12 BLUE PAISLEY TISSUES   -24  
...                                  ...  ...  
540449   ZINC T-LIGHT HOLDER STARS SMALL  -11  
541541                            Manual   -1  
541715        VICTORIAN SEWING BOX LARGE   -5  
541716  HANGING HEART JAR T-LIGHT HOLDER   -1  
541717     36 PENCILS TUBE RED RETROSPOT   -5  

[9086 rows x 8 columns]

In [199]:
def search_letter_C(the_string):
    if re.search('[C]', the_string) is None:
        return False
    return True

df[df['BasketID'].apply(search_letter_C)]

BasketID      BasketDate    Sale  CustomerID CustomerCountry  ProdID  \
141     C536379  01/12/10 09:41    27,5     14527.0  United Kingdom       D   
154     C536383  01/12/10 09:49    4,65     15311.0  United Kingdom  35004C   
235     C536391  01/12/10 10:24    1,65     17548.0  United Kingdom   22556   
236     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21984   
237     C536391  01/12/10 10:24    0,29     17548.0  United Kingdom   21983   
...         ...             ...     ...         ...             ...     ...   
540449  C581490  09/12/11 09:57    0,83     14397.0  United Kingdom   23144   
541541  C581499  09/12/11 10:28  224,69     15498.0  United Kingdom       M   
541715  C581568  09/12/11 11:57   10,95     15311.0  United Kingdom   21258   
541716  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   84978   
541717  C581569  09/12/11 11:58    1,25     17315.0  United Kingdom   20979   

                               ProdDescr  Qta  
141                             Discount   -1  
154      SET OF 3 COLOURED  FLYING DUCKS   -1  
235       PLASTERS IN TIN CIRCUS PARADE   -12  
236     PACK OF 12 PINK PAISLEY TISSUES   -24  
237     PACK OF 12 BLUE PAISLEY TISSUES   -24  
...                                  ...  ...  
540449   ZINC T-LIGHT HOLDER STARS SMALL  -11  
541541                            Manual   -1  
541715        VICTORIAN SEWING BOX LARGE   -5  
541716  HANGING HEART JAR T-LIGHT HOLDER   -1  
541717     36 PENCILS TUBE RED RETROSPOT   -5  

[9084 rows x 8 columns]

In [200]:
def search_letters_noC(the_string):
    if re.search('[a-zA-Z]', the_string) is None:
        return False
    elif re.search('[C]', the_string) is None:
        return True
    return False

df[df['BasketID'].apply(search_letters_noC)]

BasketID      BasketDate       Sale  CustomerID CustomerCountry ProdID  \
299983  A563186  12/08/11 14:51  -11062,06         NaN  United Kingdom      B   
299984  A563187  12/08/11 14:52  -11062,06         NaN  United Kingdom      B   

              ProdDescr  Qta  
299983  Adjust bad debt    1  
299984  Adjust bad debt    1

In [201]:
tmp = df[df['BasketID'].apply(lambda s: not search_letters(s))]
df = tmp.copy()
df

BasketID      BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0        536365  01/12/10 08:26  2,55     17850.0  United Kingdom  85123A   
1        536365  01/12/10 08:26  3,39     17850.0  United Kingdom   71053   
2        536365  01/12/10 08:26  2,75     17850.0  United Kingdom  84406B   
3        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029G   
4        536365  01/12/10 08:26  3,39     17850.0  United Kingdom  84029E   
...         ...             ...   ...         ...             ...     ...   
541905   581587  09/12/11 12:50   2,1     12680.0          France   22899   
541906   581587  09/12/11 12:50  4,15     12680.0          France   23254   
541907   581587  09/12/11 12:50  4,15     12680.0          France   23255   
541908   581587  09/12/11 12:50  4,95     12680.0          France   22138   
541909   581587  09/12/11 12:50    18     12680.0          France    POST   

                                  ProdDescr  Qta  
0        WHITE HANGING HEART T-LIGHT HOLDER    6  
1                       WHITE METAL LANTERN    6  
2            CREAM CUPID HEARTS COAT HANGER    8  
3       KNITTED UNION FLAG HOT WATER BOTTLE    6  
4            RED WOOLLY HOTTIE WHITE HEART.    6  
...                                     ...  ...  
541905         CHILDREN'S APRON DOLLY GIRL     6  
541906        CHILDRENS CUTLERY DOLLY GIRL     4  
541907      CHILDRENS CUTLERY CIRCUS PARADE    4  
541908        BAKING SET 9 PIECE RETROSPOT     3  
541909                              POSTAGE    1  

[462824 rows x 8 columns]

Abbiamo eliminato 9086 righe (tutte quelle che hanno la lettera C o A davanti a BasketID).

In [202]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 462824 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         462824 non-null  object 
 1   BasketDate       462824 non-null  object 
 2   Sale             462824 non-null  object 
 3   CustomerID       397925 non-null  float64
 4   CustomerCountry  462824 non-null  object 
 5   ProdID           462824 non-null  object 
 6   ProdDescr        462071 non-null  object 
 7   Qta              462824 non-null  int64  
dtypes: float64(1), int64(1), object(6)
memory usage: 31.8+ MB


Trasformo BasketID in intero.

In [203]:
df['BasketID'] = df['BasketID'].astype(int)

In [204]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 462824 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   BasketID         462824 non-null  int64  
 1   BasketDate       462824 non-null  object 
 2   Sale             462824 non-null  object 
 3   CustomerID       397925 non-null  float64
 4   CustomerCountry  462824 non-null  object 
 5   ProdID           462824 non-null  object 
 6   ProdDescr        462071 non-null  object 
 7   Qta              462824 non-null  int64  
dtypes: float64(1), int64(2), object(5)
memory usage: 31.8+ MB
